File: data/scripts/hook/external_input.lua
--
-- Simple hook-script for allowing external input drivers
--
-- The default connection-point name is extio_1, extio_2 and
-- so on. In order to activate additional ones,
--
local connpoint_handler

local function open_connpoint(prefix, ind)
-- keep an external counter so that multiple -Hexternal_input can
-- cooperate and create extio_1, extio_2 etc.
	if not ind then
		ind = _G["_external_input_index"]
		_G["_external_input_index"] = _G["_external_input_index"] + 1
	end

	local key = prefix .. "_" .. tonumber(ind);
	local vid = target_alloc(key,
	function(source, status, iotbl)
		if status.kind == "terminated" then
			delete_image(source)
			open_connpoint(prefix, ind)
		elseif status.kind == "input" then
			local fun = _G[APPLID .. "_input"]
			if (fun) then
				fun(iotbl)
			end
		end
	end
	)

	if not valid_vid(vid) then
		warning("builtin/external_input: could not open '" .. key .. "'")
	end
end

if not _G["_external_input_index"] then
	_G["_external_input_index"] = 1
end

-- allow some database control over what the name of the connection
-- point would become, and a fallback default if that is not present
local prefix = get_key("ext_io");
if not prefix then
	prefix = "extio"
end

open_connpoint(prefix)

File: data/scripts/hook/external_input.zig

var connpoint_handler = undefined;
fn open_connpoint(prefix: []const u8, ind: anytype) void {
    if (!ind) {
        ind = _G["_external_input_index"];
        _G["_external_input_index"] = _G["_external_input_index"] + 1;
    }
    var key = prefix ++ ("_" ++ tonumber(ind));
    var vid = target_alloc(key, struct { fn anon(source: anytype, status: anytype, iotbl: anytype) void {
        if (status.kind == "terminated") {
            delete_image(source);
            open_connpoint(prefix, ind);
        } else if (status.kind == "input") {
            var fun = _G[APPLID ++ "_input"];
            if (fun) {
                fun(iotbl);
            }
        }
    } }.anon);
    if (!valid_vid(vid)) {
        warning("builtin/external_input: could not open '" ++ (key ++ "'"));
    }
}

pub fn __init() void {
    if (!_G["_external_input_index"]) {
        _G["_external_input_index"] = 1;
    }
    var prefix = get_key("ext_io");
    if (!prefix) {
        prefix = "extio";
    }
    open_connpoint(prefix);
}
